### PR TITLE
Added checks while reading prelogin for invalid data

### PR DESCRIPTION
**Pull request by dthaluru**
Currently the driver panic If the server sends bad prelogin connection data.
Added additional checks and unit tests to verify the behavior.

https://github.com/microsoft/go-mssqldb/issues/64

### CODE REVIEW

**Comment by dthaluru** [comment]
> 

@microsoft-github-policy-service agree [company="Google"]

**Comment by dthaluru** [comment]
@microsoft-github-policy-service agree
@microsoft-github-policy-service agree company="Google"

File: bad_server_test.go [hunk 1]
```diff
@@ -119,6 +119,42 @@
 	})
 }
 
+func TestBadServerPreLoginPacketWithInvalidOptions(t *testing.T) {
+	testBadServer(t, func(conn net.Conn) {
+		buf := newTdsBuffer(defaultPacketSize, conn)
+		preloginPacket := []byte{4, 1, 0, 32, 0, 0, 1}
+
+		_, err = buf.Write(preloginPacket)
```
Comment (shueybubbles) [inline bad_server_test.go]: >r =  [](http://example.com/codeflow?start=7&length=4)

:=

File: bad_server_test.go [hunk 2]
```diff
@@ -119,6 +119,42 @@
...
+		}
+	})
+}
+
+func TestBadServerPreLoginPacketWithNoTerminator(t *testing.T) {
+	testBadServer(t, func(conn net.Conn) {
+		buf := newTdsBuffer(defaultPacketSize, conn)
+		preloginPacket := []byte{4, 1, 0, 32, 0, 0, 1, 0, 0, 0, 16, 0, 6, 1, 0, 22, 0, 1, 6, 0, 23, 0, 1, 0, 12, 0, 7, 208, 0, 0, 2, 1}
+
+		_, err = buf.Write(preloginPacket)
```
Comment (shueybubbles) [inline bad_server_test.go]: >= [](http://example.com/codeflow?start=9&length=1)

:=

File: bad_server_test.go [hunk 3]
```diff
@@ -119,6 +119,42 @@
...
+		}
+	})
+}
+
+func TestBadServerPreLoginPacketWithMissingData(t *testing.T) {
+	testBadServer(t, func(conn net.Conn) {
+		buf := newTdsBuffer(defaultPacketSize, conn)
+		preloginPacket := []byte{4, 1, 0, 32, 0, 0, 1, 0, 0, 0, 16, 0, 6, 1, 0, 22, 0, 1, 6, 0, 23, 0, 1, 255, 12, 0, 7, 208, 0, 0, 2, 1}
+
+		_, err = buf.Write(preloginPacket)
```
Comment (shueybubbles) [inline bad_server_test.go]: >=  [](http://example.com/codeflow?start=9&length=2)

:=

**Review by shueybubbles — CHANGES_REQUESTED** [review]
:clock1:

**Review by shueybubbles — APPROVED** [review]
thx for the contribution!

**Comment by dthaluru** [comment]
@shueybubbles  I removed the @apoorvdeshmukh and @stuartpa from the review request by accident and not able to add them back. Can you please add them back? Thanks!!

**Pull request merged**